scene/gui/text_edit.cpp

```diff
@@ -2588,118 +2588,110 @@ void TextEdit::_do_backspace(bool p_word, bool p_all_to_left) {
+			// Remove overlapping carets.
+			for (int j = i + 1; j < caret_edit_order.size(); j++) {
+				// If a selection started from a different line and ended to this line, change prev_line to the selection's beginning.
+				if (has_selection(caret_edit_order[j]) && get_selection_to_line(caret_edit_order[j]) == get_caret_line(caret_idx) && get_selection_from_line(caret_edit_order[j]) < get_caret_line(caret_idx)) {
+					prev_line = get_selection_from_line(caret_edit_order[j]);
+				} else if (get_caret_line(caret_edit_order[j]) != get_caret_line(caret_idx)) {
 					break;
 				}
 
-				deselect(caret_edit_order[j]);
 				carets_to_remove.push_back(caret_edit_order[j]);
-				set_caret_column(0, caret_idx == 0, caret_idx);
-				i = j;
+				i = j; // Don't update caret_idx, just mark the next caret from which to continue the operation.
 			}
-			continue;
-		}
 
-		if (p_word) {
-			// Save here as the caret may change when resolving overlaps.
-			int from_column = get_caret_column(caret_idx);
-			int column = get_caret_column(caret_idx);
+		} else if (p_word && get_caret_column(caret_idx) > 0) {
+			if (has_selection(caret_idx)) {
+				delete_selection(caret_idx);
+				continue;
+			}
+
+			// Delete word to the left of the caret.
+			prev_line = get_caret_line(caret_idx);
+			prev_column = get_caret_column(caret_idx);
+
 			// Check for the case "<word><space><caret>" and ignore the space.
-			// No need to check for column being 0 since it is checked above.
 			if (is_whitespace(text[get_caret_line(caret_idx)][get_caret_column(caret_idx) - 1])) {
-				column -= 1;
+				prev_column -= 1;
 			}
-			// Get a list with the indices of the word bounds of the given text line.
-			const PackedInt32Array words = TS->shaped_text_get_word_breaks(text.get_line_data(get_caret_line(caret_idx))->get_rid());
-			if (words.is_empty() || column <= words[0]) {
-				// If "words" is empty, meaning no words are left, we can remove everything until the beginning of the line.
-				column = 0;
+
+			PackedInt32Array words = TS->shaped_text_get_word_breaks(text.get_line_data(get_caret_line(caret_idx))->get_rid());
+			if (words.is_empty() || prev_column <= words[0]) {
+				// No words to the left: We can remove everything until the beginning of the line.
+				prev_column = 0;
 			} else {
-				// Otherwise search for the first word break that is smaller than the index from we're currently deleting.
-				for (int c = words.size() - 2; c >= 0; c = c - 2) {
-					if (words[c] < column) {
-						column = words[c];
+				for (int j = words.size() - 2; j >= 0; j -= 2) {
+					if (words[j] < prev_column) {
+						prev_column = words[j];
 						break;
 					}
 				}
 			}
 
-			// Check for any other carets in this range.
-			int overlapping_caret_index = -1;
+			// Remove overlapping carets.
 			for (int j = i + 1; j < caret_edit_order.size(); j++) {
```

Comment:
Given this overlapping check now looks like the other one, would be nice to find a way to avoid duplicating the logic.

Reply:
They aren't identical, so I'll have to use conditions in the middle, but sure.